uswds/uswds-site
# USWDS-Site: Implement "copy code" button component previews

# Summary

Adds a button to quickly copy component code to the clipboard instead of requiring users to click and drag. 

## Related issue

Closes #2055 

## Preview link

Preview link:
[Components →](https://federalist-ead78f8d-8948-417c-a957-c21ec5617a57.sites.pages.cloud.gov/preview/uswds/uswds-site/cm-copy-component-btn/components/overview/)
[Templates →](https://federalist-ead78f8d-8948-417c-a957-c21ec5617a57.sites.pages.cloud.gov/preview/uswds/uswds-site/cm-copy-component-btn/templates/)

## Problem statement

In its current state, USWDS Site requires users to click and drag to select component code they wish to copy and implement on their websites. This can be a painful experience with larger code blocks or our template examples. 

## Solution

Implement a button within `component code` accordions to allow users to quickly copy component code to their clipboard without having to click and drag

This is a common pattern seen on most websites with code snippits.

For example, GitHub offers this when users hover a code block:
```
This is a code block example!
Hover over me to view the copy code button  →
```

## Testing and review

1. Visit any Component or Template page
2. Open `Component Code` accordion
3. View button in multiple window sizes
4. Scroll to view sticky state
5. Activate button with mouse click or keyboard navigation
6. Inspect success state
7. Paste code in text editor and verify it was correctly copied
8. Return to the code example and test with screen readers on Chrome, Firefox, **and** Safari

---

### Further considerations

#### Refactoring as a USWDS component
This is a very basic implementation of this button using the tools provided by USWDS and some javascript. James' and I briefly discussed that the minimum requirements would be to implement this as a USWDS component rather than just a 
feature on our site

#### Redesign component code section?
Currently, our components with multiple variants are all grouped together into once code block. With the copy code button, these means that all variants will be copied to the clipboard together and the user will have to remove them manually in their text editor.

Comparatively, in it's current state users have to visually determine which part of the code they wish to copy and precisely select only the block they want before bringing it to their browser, or copy all code manually and do the same as above.

It may be a good idea to revisit how these snippets are displayed on the page. Instead of grouping them into one large block, would it be better to give each variant a header within the accordion, and its own code block?

**EXAMPLE**

<details>
   <summary>Component Code</summary>

**Variant 1**
```
Variant 1 code snippit here with its own copy button
```

**Variant 2**
```
Variant 2 code snippit here with its own copy button
```

</details>


#### Text vs Icon
Do we like the icons with screen-reader only text? Would we prefer to have text buttons on either both states or even mobile only?

#### Appear on hover
Should be button only be visable once users hover the code block, similar to the GH example above?
